python/tank/bootstrap/manager.py

```diff
@@ -646,16 +649,23 @@ def _cache_apps(self, tk, progress_callback, do_post_install=False):
         descriptors = []
         for env_name in pc.get_environments():
 
-            env_obj = pc.get_environment(env_name)
+            # Select the site or project environment based on the configuration project id.
+            if (not config_project_id and env_name == "site") or (config_project_id and env_name != "site"):
```

Comment:
Done, with a call to `PICK_ENVIRONMENT_CORE_HOOK_NAME` and fallback call to `get_environments()`.